static/app/api.tsx

```diff
@@ -180,85 +101,64 @@ export const initApiClientErrorHandling = () =>
-  fullUrl = resolveHostname(fullUrl, options.host);
-
-  // Append query parameters
-  if (params) {
-    fullUrl += fullUrl.includes('?') ? `&${params}` : `?${params}`;
-  }
-
-  return fullUrl;
-}
-
-/**
- * Check if the API response says project has been renamed.  If so, redirect
- * user to new project slug
- */
-// TODO(ts): refine this type later
-export function hasProjectBeenRenamed(response: ResponseMeta) {
-  const code = response?.responseJSON?.detail?.code;
-
-  // XXX(billy): This actually will never happen because we can't intercept the 302
-  // jQuery ajax will follow the redirect by default...
-  //
-  // TODO(epurkhiser): We use fetch now, is the above comment still true?
-  if (code !== PROJECT_MOVED) {
-    return false;
-  }
-
-  const slug = response?.responseJSON?.detail?.extra?.slug;
+export type ApiResult<Data = any> = [
+  data: Data,
+  statusText: string | undefined,
+  resp: ResponseMeta | undefined,
+];
 
-  redirectToProject(slug);
-  return true;
-}
+export type ResponseMeta = {
+  /**
+   * Get a header value from the response
+   */
+  getResponseHeader: (header: string) => string | null;
+  /**
+   * The response body decoded from json
+   */
+  responseJSON: Record<string, any> | undefined;
+  /**
+   * The string value of the response
+   */
+  responseText: string;
+  /**
+   * The response status code
+   */
+  status: Response['status'];
+  /**
+   * The response status code text
+   */
+  statusText: Response['statusText'];
+};
 
-// TODO(ts): move this somewhere
 export type APIRequestMethod = 'POST' | 'GET' | 'DELETE' | 'PUT';
-
-type FunctionCallback<Args extends any[] = any[]> = (...args: Args) => void;
-
-export type RequestCallbacks = {
+export type RequestCallbackOptions = {
   /**
    * Callback for the request completing (success or error)
    */
-  complete?: (resp: ResponseMeta, textStatus: string) => void;
+  complete?: (responseMeta: ResponseMeta, statusText: string) => void;
   /**
    * Callback for the request failing with an error
    */
-  // TODO(ts): Update this when sentry is mostly migrated to TS
-  error?: FunctionCallback;
+  error?: (response: ResponseMeta, statusText: string, errorThrown: string) => void;
   /**
    * Callback for the request completing successfully
    */
-  success?: (data: any, textStatus?: string, resp?: ResponseMeta) => void;
+  success?: (responseData: any, statusText: string, responseMeta: ResponseMeta) => void;
 };
 
-export type RequestOptions = RequestCallbacks & {
+export interface RequestOptions extends RequestCallbackOptions {
   /**
    * Set true, if an authentication required error is allowed for
    * a request.
    */
   allowAuthError?: boolean;
+  /**
+   * By default, requests will be aborted anytime api.clear() is called,
+   * which is commonly used on unmounts. When cancelable is set to false, the
+   * request is opted out of this behavior. Useful for when you still
+   * want to cache a request on unmount.
+   */
+  cancelable?: boolean;
```

Comment:
note: might be out of scope, but react-query handles this under the hood automatically as soon as you opt-in to using the passed `AbortSignal` of the `QueryFunctionContext`:

https://tanstack.com/query/v5/docs/framework/react/guides/query-cancellation

The defaults are the opposite though - it’s not canceled so that it can be cached for later, and cancelation is opt-in.

I still think it might be worth letting this be handled by the lib, as the “unmount” (= `useEffect` cleanup function) event is component scoped: Suppose one query is used by two components, and one of them unmounts: This will lead to the request being canceled, and the other component will likely not be able to display anything.

Query handles this by only aborting after all usages have been unmounted.

Reply:
I agree, I ultimately want rq to handle this for us, but it's going to snowball into a massive change if I do that right now 😓.

There is another reason where this can be an issue btw, which is that if the instance of the API client is passed around, it is currently possible for it to end up canceling requests that belong to the children or parent components. I dont think it ever happens, but the danger is there, and it would result in a particularly nasty issue to track down

Reply:
yeah let’s make a separate issue for this. The problem is likely that not everything is using react-query yet, so we’re kinda dependent on that migration.